File: cpp/grade-school/grade_school.cpp
#include "grade_school.h"
#include <algorithm>

namespace grade_school {

bool school::add(const std::string& name, int grade) {
    // Check if the student already exists in any grade
    for (auto& [grade_num, students] : school_roster) {
        if (std::find(students.begin(), students.end(), name) != students.end()) {
            return false; // Student already exists
        }
    }
    
    // Add the student to the specified grade
    school_roster[grade].push_back(name);
    
    // Sort the students in the grade alphabetically
    std::sort(school_roster[grade].begin(), school_roster[grade].end());
    
    return true;
}

std::vector<std::string> school::grade(int grade) const {
    // If the grade exists, return its students, otherwise return an empty vector
    if (school_roster.find(grade) != school_roster.end()) {
        return school_roster.at(grade);
    }
    return {};
}

std::map<int, std::vector<std::string>> school::roster() const {
    return school_roster;
}

}  // namespace grade_school

File: python/grade-school/grade_school.py
class School:
    def __init__(self):
        self._students = {}  # Dictionary to store students by grade
        self._addition_results = []  # List to track addition results

    def add_student(self, name, grade):
        # Check if student already exists in any grade
        for grade_students in self._students.values():
            if name in grade_students:
                self._addition_results.append(False)
                return False
        
        # Add student to the specified grade
        if grade not in self._students:
            self._students[grade] = []
        
        self._students[grade].append(name)
        # Sort students in this grade alphabetically
        self._students[grade].sort()
        
        self._addition_results.append(True)
        return True

    def roster(self):
        result = []
        # Get all grades and sort them
        sorted_grades = sorted(self._students.keys())
        
        # Add students from each grade in order
        for grade in sorted_grades:
            result.extend(self._students[grade])
        
        return result

    def grade(self, grade_number):
        # Return an empty list if the grade doesn't exist
        if grade_number not in self._students:
            return []
        
        # Return a copy of the list to prevent modification
        return self._students[grade_number].copy()

    def added(self):
        return self._addition_results
